### ISSUE TITLE
Unpooling with torch.autograd

### ISSUE THREAD

**Issue by ankurhanda** [twitter/torch-autograd#50]
I'm max pooling and saving the indices of max values. I'd like to use these indices when I'm upsampling later in the network. However, Unpooling layer doesn't have any parameters and whenever I run the program I get this error 

"A node type was not returned. This is either because a gradient was not defined, or the input is independent of the output" 

Could you suggest where exactly I'd need to make changes? Here's the code 

```
-- Define a standard nn model:
require 'nn'
t = require 'torch'
grad = require 'autograd'

local model = nn.Sequential()
model:add(nn.SpatialConvolutionMM(1, 1, 3, 3, 1, 1, 1, 1))
local pool1 = nn.SpatialMaxPooling(2,2)
local idx = pool1.indices
model:add(pool1)

-- Note that this model could have been pre-trained, and reloaded from disk.

-- Functionalize the model:
local modelf, params = grad.functionalize(model)

-- The model can now be used as part of a regular autograd function:
local loss = grad.nn.MSECriterion()

function Unpool(h1,idx)
	
	local pred = t.Tensor(1,4,4):zero()

	D,H,W = h1:size(1),h1:size(2),h1:size(3)

	for d=1,D do
		for h=1,H do
			for w=1,W do
				pred[d][h*2][w*2] = h1[d][h][w]
		---		idx_h = math.floor(idx[d][h][w]/4)+1
		---		idx_w = math.fmod(idx[d][h][w],4)
		---		pred[d][idx_h][idx_w] = h[d][h][w]
			end
		end
	end	

	return pred
end

neuralNet = function(params, x, y)
   local h1 = modelf(params, x)
   local pred= Unpool(h1,idx)
   return loss(pred, y)
end

-- gradients:
dneuralNet = grad(neuralNet)

-- some data:
x = t.randn(1,4,4)
y = t.Tensor(1,4,4):zero() 

-- compute loss and gradients wrt all parameters in params:
dparams, loss = dneuralNet(params,x, y)

print(loss)


```

**Comment by ankurhanda** [comment]
Just checked AutoModule might be relevant but how to instantiate a function with no arguments is not clear to me.

**Comment by alexbw** [comment]
Hey, right now we don't return the gradients of the max values. This is a
good use case and we should support it. I'll work on a patch.

**Comment by alexbw** [comment]
@nkoumchatzky any advice here? I don't think that this use of pulling out object member values is supported by autograd...

**Comment by nkoumchatzky** [comment]
I can't run it now as I don't have my laptop with me but looking Arthur
code it seems that the Unpooling shouldn't have a params argument at all,
am I correct?

Nico

**Comment by ankurhanda** [comment]
@nkoumchatzky Yes, Unpooling shouldn't have any parameters at all --- I had them in the function only to probe autograd's internal way of handling functions whereby default the parameters have to be passed to the function. It would be great to figure out how to do this properly. 

The last resort then is to write an Unpooling.lua file within torch and compile it against the main torch code. Doing that way I will have to define updateGradInput and accGrad etc. which is what you shouldn't have to if you were using autograd.

**Comment by alexbw** [comment]
There's actually several issues with the code above — we don't currently support assignment in `autograd`'d functions, which you do here: `pred[d][h*2][w*2] = h1[d][h][w]`

And also, you grab a reference to `pool1.indices` as an upvalue, and unfortunately we can't give a guarantee that autograd, when using `pool1` module inside of a functionalized model, will actually modify the specific reference you have. We should have better error messages here. 

@luketwitter, Can we do any better on the error-reporting front here? I'm not sure how/if we want to do this.

**Comment by thouis** [comment]
I'm curious about the right way to do unpool with indices, as well.  I'm hoping to reproduce some work from this ICCV paper: http://arxiv.org/abs/1505.04366

Is there a workaround, or general sketch of how to create one?  (let me know if I should file that question as a separate issue.)

**Comment by alexbw** [comment]
This is the place to post this. Right now, we don't support function with
multiple return values in autograd, so

val, index = torch.max(A)

will only return

val = torch.max(A)

We should probably implement a numpy-style

index = torch.argmax(A)

in order to do this elegantly.

@clementfarabet, what do you think?

**Comment by fmassa** [comment]
@thouis there are some implementations of unpooling that could be interesting to you.
check for example https://gist.github.com/andreaskoepf/2539cf7493dfe032ced3 and the recent PR https://github.com/torch/nn/pull/520

**Comment by alexbw** [comment]
Oh very nice. It's extremely easy to wrap nn modules, so my bet is that this would work out of the box.

**Comment by thouis** [comment]
I installed nn with PR torch/nn#520 included, and grad wraps nn.SpatialMaxUnpooling automatically, but grad.nn.SpatialMaxPooling / grad.nn.SpatialMaxUnpooling don't work in that form, as the latter wants to dig into the former to get indices. 

Trying to use the versions in nn doesn't seem to work either, giving  "`bad argument #2 to 'SpatialMaxPooling_updateOutput' (torch.FloatTensor expected, got userdata)`"

I'll experiment with the other approach.

**Comment by clement-masson** [comment]
Hi.
@thouis 
>Trying to use the versions in nn doesn't seem to work either, giving "bad argument #2 to 'SpatialMaxPooling_updateOutput' (torch.FloatTensor expected, got userdata)"
Can you give more details about this error ? 
What line is it ? Does it occur only with float tensors ?

**Comment by thouis** [comment]
I've put test code in https://gist.github.com/thouis/fd50b8c28efcc078765e

I get this error:
```
luajit: ...uis/torch/install/share/lua/5.1/nn/SpatialMaxPooling.lua:36: bad argument #2 to 'SpatialMaxPooling_updateOutput' (torch.FloatTensor expected, got userdata)
stack traceback:
	[C]: in function 'SpatialMaxPooling_updateOutput'
	...uis/torch/install/share/lua/5.1/nn/SpatialMaxPooling.lua:36: in function 'forward'
	/Users/thouis/torch/install/share/lua/5.1/nn/Module.lua:276: in function </Users/thouis/torch/install/share/lua/5.1/nn/Module.lua:275>
	[C]: in function 'pooler'
	test_maxunpool.lua:14: in function 'fun'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:19: in function 'funOnly'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:114: in function 'df'
	test_maxunpool.lua:21: in main chunk
	[C]: in function 'dofile'
	...ouis/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x010e651bd0
```

My installation may be broken (I had to change the tensor type for a few fields for it to work, even before calling grad()), so I'm going to start over and rebuild from scratch.

**Comment by clement-masson** [comment]
@thouis 
Strange. When I remove the lines related to autograd (I do not have installed it), your code works on my installation. It may indeed be that yours is broken.

**Comment by thouis** [comment]
@fmassa can probably clarify, but I think this may be something that needs to be fixed on the autograd side, not yours.  If the changes for SpatialMaxUnpooling get merged, I think autograd will need a different wrapper for it and SpatialMaxPooling (to expose the indices).

**Comment by alexbw** [comment]
Is this still an issue?

**Comment by thouis** [comment]
When I tried it a few days ago, I was still unable to get it to work.

**Comment by alexbw** [comment]
This to me looks like a type-casting issue. I commented on your gist, you should cast the module, not its member tensors. Also, be sure to pull the most recent autograd, which has changes to the `functionalize` API which allows the use of `nn` modules. Happy to to talk in gitter, as well.

**Comment by alexbw** [comment]
Hey @thouis, I'm going to close this. I think the best way to open this back up (if you're still interested in having your code snippet working) is with a pull request containing a failing test after you use `util.cast` to take care of what I believe are type issues with your code.
